pkg/controller/controller.go

```diff
@@ -352,13 +364,104 @@ func (s *Service) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest
+					// Having a protection policy associated with the vg makes it invalid for metro replication.
+					if vg.ProtectionPolicyID != "" {
+						return nil, status.Errorf(codes.FailedPrecondition,
+							"volume group %s has a protection policy assigned making it incompatible for usage with metro replication.", vg.Name)
+					}
+
+					// Confirm VG is part of a metro session and pause the session to allow adding a volume to the vg
+					if vg.MetroReplicationSessionID != "" {
+						// make sure session is in good state
+						metroVGSession, err := arr.Client.GetReplicationSessionByID(ctx, vg.MetroReplicationSessionID)
+						if metroVGSession.State != gopowerstore.RsStateOk {
+							return nil, status.Errorf(codes.FailedPrecondition,
+								"cannot add volumes to volume group %s because the metro replication session is not in running state.", vg.Name)
+						}
+
+						// if session is running, pause metro
+						log.Debugf("metro session state is OK. Pausing metro volume group session for %s", vg.Name)
+						_, err = arr.Client.ExecuteActionOnReplicationSession(ctx, vg.MetroReplicationSessionID, gopowerstore.RsActionPause, nil)
+						if err != nil {
+							return nil, status.Errorf(codes.Internal, "unable to pause metro replication session on volume group %s: %s", vg.Name, err.Error())
+						}
+
+						// Metro session paused. Volume will be added to the vg and metro will be resumed below.
+						startMetroReplicationSession = resumeMetroVolumeGroupSession
+					} else {
+						// vg exists but is not yet metro replicated
+						// if it already has volumes attached, exit with error
+						if len(vg.Volumes) > 0 {
+							return nil, status.Errorf(codes.FailedPrecondition,
+								"volume group %s found with volumes attached, but not part of a metro replication session.", vg.Name)
+						}
+
+						// vg needs to be write-order consistent for metro replication
+						if !vg.IsWriteOrderConsistent {
+							return nil, status.Errorf(codes.FailedPrecondition,
+								"volume group %s is not write-order consistent and cannot be used for metro replication.", vg.Name)
+						}
+
+						log.Infof("reusing empty volume group %s for metro replication", vg.Name)
+
+						// configure the metro session after adding the new volume
+						startMetroReplicationSession = configureMetroVolumeGroupSession
+					}
+				} else if apiError, ok := err.(gopowerstore.APIError); ok && apiError.NotFound() {
+					// VG does not yet exist. Create it on powerstore array
+					log.Infof("creating volume group %s for metro replication", vgName)
+					newVG, err := arr.Client.CreateVolumeGroup(ctx, &gopowerstore.VolumeGroupCreate{
+						Name:                   vgName,
+						IsWriteOrderConsistent: true,
+					})
+					if err != nil {
+						return nil, status.Errorf(codes.Internal, "unable to create volume group %s on PowerStore array: %s", vgName, err.Error())
+					}
+
+					vg, err = arr.Client.GetVolumeGroup(ctx, newVG.ID)
```

Comment:
We can avoid this GetVG(). The expectation is that CreateVG() should have successfully created it, and we error handle the Create call.

Reply:
`CreateVG()` will only return the volume group ID.
I'm calling querying for the vg here so when we pass it to the function that starts metro replication, it will have all necessary parameters to get the job done.